webcrypto/subtle_crypto.go

```diff
@@ -137,8 +141,87 @@ func (sc *SubtleCrypto) Verify(
 //   - SHA-512
 //
 // The `data` parameter should contain the data to be digested.
-func (sc *SubtleCrypto) Digest(algorithm goja.Value, data interface{}) *goja.Promise {
-	return nil
+func (sc *SubtleCrypto) Digest(algorithm interface{}, data goja.Value) *goja.Promise {
+	promise, resolve, reject := sc.makeHandledPromise()
+
+	// Validate that the value we received is either an ArrayBuffer, TypedArray, or DataView
+	// This uses the technique described in https://github.com/dop251/goja/issues/379#issuecomment-1164441879
+	var (
+		isArrayBuffer = IsInstanceOf(sc.vu.Runtime(), data, ArrayBufferConstructor)
+		isDataView    = IsInstanceOf(sc.vu.Runtime(), data, DataViewConstructor)
+		isTypedArray  = IsTypedArray(sc.vu.Runtime(), data)
+	)
+
+	if !isArrayBuffer && !isDataView && !isTypedArray {
+		reject(fmt.Errorf("data must be an ArrayBuffer, TypedArray, or DataView"))
+		return promise
+	}
+
+	// 2.
+	// Cast the data to a Goja Object, and, as we're now sure it's
+	// either an ArrayBuffer, or a view on an ArrayBuffer, we can
+	// get the underlying ArrayBuffer by exporting its buffer property
+	// to a Goja ArrayBuffer, and then getting its underlying Go slice
+	// by calling the `Bytes()` method.
+	//
+	// Doing so conviniently also copies the underlying buffer, which
+	// is required by the specification.
+	// See https://www.w3.org/TR/WebCryptoAPI/#SubtleCrypto-method-digest
+	asObject := data.ToObject(sc.vu.Runtime())
+	arrayBuffer, ok := asObject.Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(fmt.Errorf("could not get ArrayBuffer from data"))
```

Comment:
I'm familiar with TypedArrays, but I was hoping we could get their ArrayBuffer using the Goja Go API, instead of doing it in JS via the `buffer` property. It would be safer if we didn't rely on duck typing :sweat_smile: If that's not possible, then I suppose it's the only thing we can do.

Though your latest change breaks passing an actual `ArrayBuffer` instance, which should be supported according to the spec, so we should type assert on both `data` and `data.ToObject(rt).Get("buffer")`.

Also, let's check for `null`/`undefined` as well.

Reply:
I see what you mean. I understand that doing it in Go is not possible now. If I remember correctly, when I experimented with it, treating the value as a `goja.ArrayBuffer` led to a behavior where the ArrayBuffer was empty.  The solution I adopted here is based on this discussion: https://github.com/dop251/goja/issues/379.

Good catch on the `ArrayBuffer` breaking change 👍🏻 I'll update it accordingly 🙇🏻